api/lib/domain/services/assessment-service.js

```diff
@@ -53,6 +75,72 @@ function selectNextChallengeId(course, currentChallengeId, assessment) {
 
 module.exports = {
 
+  _getDifficultyOfKnowledge,
+  _nextNode,
+  _propagateKnowledge,
+
+  populateScore(assessment, answers, knowledgeData) {
```

Comment:
Woy ! elle fait mal celle-là :-)

1. mettre des tests unitaires : t'es dans un service, t'as la main sur tout. au pire tu stub / mock.
2. une fois que tu as tes tests, découper ta méthode en sous-méthodes bien nommées
3. refactorer : avoir une seule façon de faire des boucles for (utiliser la méthode standard `Array#forEach()`, remplacer `if (challenge !== 'undefined')` par `if (challenge)` directement